cpp/src/decisiontree/batched-levelalgo/kernels.cuh

```diff
@@ -362,50 +367,126 @@ __global__ void computeSplitClassificationKernel(
     col = select(colIndex, treeid, node.info.unique_id, seed, input.N);
   }
 
-  for (IdxT i = threadIdx.x; i < len; i += blockDim.x) shist[i] = 0;
+  // populating shared memory with initial values
+  for (IdxT i = threadIdx.x; i < pdf_shist_len; i += blockDim.x) pdf_shist[i] = 0;
+  for (IdxT j = threadIdx.x; j < cdf_shist_len; j += blockDim.x) cdf_shist[j] = 0;
   for (IdxT b = threadIdx.x; b < nbins; b += blockDim.x)
     sbins[b] = input.quantiles[col * nbins + b];
+
+  // synchronizing above changes across block
   __syncthreads();
+
+  // compute pdf shared histogram for all bins for all classes in shared mem
   auto coloffset = col * input.M;
-  // compute class histogram for all bins for all classes in shared mem
   for (auto i = range_start + tid; i < end; i += stride) {
+    // each thread works over a data point and strides to the next
     auto row = input.rowids[i];
     auto d = input.data[row + coloffset];
     auto label = input.labels[row];
+    // breakflag to check if data is greater than largest bin
+    bool breakflag = false;
     for (IdxT b = 0; b < nbins; ++b) {
-      auto isRight = d > sbins[b];  // no divergence
-      auto offset = b * 2 * nclasses + isRight * nclasses + label;
-      atomicAdd(shist + offset, 1);  // class hist
+      if(d <= sbins[b]) { // shist (0 -> nbins*nclasses - 1)
+        auto offset = b * nclasses + label;
+        atomicAdd(pdf_shist + offset, 1);
+        breakflag = true;
+        break;
+      }
     }
+    // case when d is larger than all bins
+    if(!breakflag) atomicAdd(pdf_shist + nbins*nclasses + label, 1);
```

Comment:
This should never happen with new way of quantile computation once #3586 gets merged. As it does not incur a minimal penalty, we can keep this for now.